numpy/core/src/multiarray/dtypemeta.c

```diff
@@ -0,0 +1,269 @@
+ * a newly created dtype subclass (DTypeMeta instances).
+ * There are some subtleties that need to be remembered when doing this,
+ * first for the class objects itself it could be either a HeapType or not.
+ * Since we are defining the DType from C, we will not make it a HeapType,
+ * thus making it identical to a typical *static* type (except that we
+ * malloc it). We could do it the other way, but there seems no reason to
+ * do so.
+ *
+ * The DType instances (the actual dtypes or descriptors), are based on
+ * prototypes which are passed in. These should not be garbage collected
+ * and thus Py_TPFLAGS_HAVE_GC is not set. (We could allow this, but than
+ * would have to allocate a new object, since the GC needs information before
+ * the actual struct).
+ *
+ * The above is the reason why we should works exactly like we would for a
+ * static type here.
+ * Otherwise, we blurry the lines between C-defined extension classes
+ * and Python subclasses. e.g. `class MyInt(int): pass` is very different
+ * from our `class Float64(np.dtype): pass`, because the latter should not
+ * be a HeapType and its instances should be exact PyArray_Descr structs.
+ *
+ * @param descr The descriptor that should be wrapped.
+ * @param name The name for the DType, if NULL the type character is used.
+ *
+ * @returns 0 on success, -1 on failure.
+ */
+NPY_NO_EXPORT int
+dtypemeta_wrap_legacy_descriptor(PyArray_Descr *descr)
+{
+    if (Py_TYPE(descr) != &PyArrayDescr_Type) {
+        PyErr_Format(PyExc_RuntimeError,
+                "During creation/wrapping of legacy DType, the original class "
+                "was not PyArrayDescr_Type (it is replaced in this step).");
+        return -1;
+    }
+
+    /*
+     * Note: we have no intention of freeing the memory again since this
+     * behaves identically to static type definition (see comment above).
+     * This is seems cleaner for the legacy API, in the new API both static
+     * and heap types are possible (some difficulty arises from the fact that
+     * these are instances of DTypeMeta and not type).
+     * In particular our own DTypes can be true static declarations.
+     * However, this function remains necessary for legacy user dtypes.
+     */
+
+    const char *scalar_name = descr->typeobj->tp_name;
+    /*
+     * We have to take only the name, and ignore the module to get
+     * a reasonable __name__, since static types are limited in this regard
+     * (this is not ideal, but not a big issue in practice).
+     * This is what Python does to print __name__ for static types.
+     */
+    const char *dot = strrchr(scalar_name, '.');
+    if (dot) {
+        scalar_name = dot + 1;
+    }
+    ssize_t name_length = strlen(scalar_name) + 14;
+
+    char *tp_name = malloc(name_length);
+    if (tp_name == NULL) {
+        PyErr_NoMemory();
+        return -1;
+    }
+
+    snprintf(tp_name, name_length, "numpy.dtype[%s]", scalar_name);
```

Comment:
True... I suppose the metaclass could probably have a `__repr__` slot to fix some of it. Although, it may be that to actually make it copy-pastable, we need a `.` in there, which may mean it can only be done by making everything HeapTypes (which is not totally unreasonable).

Reply:
Can we do something like numpy.dtype_%s ? Although this may be a bit ugly and little less self-explanatory of it being a subclass of np.dtype.

Reply:
We could do various things I guess, but it will be a while before we should seriously consider exposing these in the API, and I think until then, it will be fine to modify the `__repr__` as we like.